python/pyarrow/array.pxi

```diff
@@ -2771,6 +2771,11 @@ cdef class ExtensionArray(Array):
         """
         return self.storage.to_numpy(**kwargs)
 
+    cdef getitem(self, int64_t i):
+        scalar = ExtensionScalar.wrap(GetResultValue(self.ap.GetScalar(i)))
+        return self.type.__arrow_ext_scalar_class__().from_storage(
+            self.type, scalar.value)
```

Comment:
@jorisvandenbossche Does this approach make sense or should we unwrap the value and wrap it into the correct subclassed ExtensionScalar? I'm sorry, I'm not very familiar with the machinery here.

Reply:
The alternative would be to update `Scalar.wrap` / `pyarrow_wrap_scalar` to do this already, similarly how this is done for Array wrapping (if you look for how `__arrow_ext_class__` gets used, you might be able to do something similar, although the scalar wrap code is structured a bit differently)

Reply:
Ah, reading further, I see you already defined a `get_scalar_class_from_type` similarly as how it is done for arrays. In that case I think it should actually be possible to update `Scalar.wrap` as well, and then the above `getitem` override shouldn't be needed?

Reply:
Done. Please check.